File: crates/wasm/examples/encrypt.ts
#!/usr/bin/env -S deno run --allow-read --allow-write

// SPDX-FileCopyrightText: 2024 Shun Sakai
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

import * as cli from "@std/cli";

import * as command from "@cliffy/command";

import * as abcrypt from "../pkg/abcrypt_wasm.js";

import { VERSION } from "./version.ts";

const { args, options } = await new command.Command()
  .name("encrypt")
  .version(VERSION)
  .description("An example of encrypting to the abcrypt encrypted data format.")
  .option("--argon2-type <TYPE:integer>", "Set the Argon2 type.", {
    default: 2,
  })
  .option("--argon2-version <VERSION:integer>", "Set the Argon2 version.", {
    default: 0x13,
  })
  .option("-m, --memory-cost <NUM:integer>", "Set the memory size in KiB.", {
    default: 19456,
  })
  .option("-t, --time-cost <NUM:integer>", "Set the number of iterations.", {
    default: 2,
  })
  .option("-p, --parallelism <NUM:integer>", "Set the degree of parallelism.", {
    default: 1,
  })
  .arguments("<INFILE:file> <OUTFILE:file>")
  .parse();

const plaintext = Deno.readFileSync(args[0]);

const passphrase = new TextEncoder()
  .encode(cli.promptSecret("Enter passphrase: ")!);
const ciphertext = abcrypt.encryptWithVersion(
  plaintext,
  passphrase,
  options.argon2Type,
  options.argon2Version,
  options.memoryCost,
  options.timeCost,
  options.parallelism,
);

Deno.writeFileSync(args[1], ciphertext);

File: crates/python/examples/encrypt.py
#!/usr/bin/env python3

# SPDX-FileCopyrightText: 2024 Shun Sakai
#
# SPDX-License-Identifier: Apache-2.0 OR MIT

import argparse
import getpass
import sys

import abcrypt_py
import version


def main() -> None:
    parser = argparse.ArgumentParser(
        prog="encrypt",
        description="An example of encrypting a file to the abcrypt encrypted data format",
    )
    parser.add_argument(
        "-V", "--version", action="version", version=version.VERSION
    )
    parser.add_argument(
        "-o",
        "--output",
        default=sys.stdout.buffer,
        type=argparse.FileType("wb"),
        help="output the result to a file",
        metavar="FILE",
    )
    parser.add_argument(
        "--argon2-type",
        default=2,
        type=int,
        choices=range(3),
        help="set the Argon2 type",
        metavar="TYPE",
    )
    parser.add_argument(
        "--argon2-version",
        default=0x13,
        type=int,
        choices=[0x10, 0x13],
        help="set the Argon2 version",
        metavar="VERSION",
    )
    parser.add_argument(
        "-m",
        "--memory-cost",
        default=19456,
        type=int,
        help="set the memory size in KiB",
        metavar="NUM",
    )
    parser.add_argument(
        "-t",
        "--time-cost",
        default=2,
        type=int,
        help="set the number of iterations",
        metavar="NUM",
    )
    parser.add_argument(
        "-p",
        "--parallelism",
        default=1,
        type=int,
        help="set the degree of parallelism",
        metavar="NUM",
    )
    parser.add_argument(
        "input",
        nargs="?",
        default=sys.stdin,
        type=argparse.FileType("rb"),
        help="input file",
        metavar="FILE",
    )
    args = parser.parse_args()

    plaintext = args.input.read()

    passphrase = bytes(getpass.getpass("Enter passphrase: "), encoding="utf-8")
    ciphertext = abcrypt_py.encrypt_with_version(
        plaintext,
        passphrase,
        args.argon2_type,
        args.argon2_version,
        args.memory_cost,
        args.time_cost,
        args.parallelism,
    )

    args.output.write(ciphertext)


if __name__ == "__main__":
    main()
